pil-analyzer/src/condenser.rs

```diff
@@ -939,6 +941,78 @@ fn shift_local_var_refs(e: &mut Expression, shift: u64) {
         .for_each(|e| shift_local_var_refs(e, shift));
 }
 
+fn try_algebraic_expression_to_expression<T: FieldElement>(
+    e: &AlgebraicExpression<T>,
+) -> Result<Expression, EvalError> {
+    Ok(match e {
+        AlgebraicExpression::Reference(AlgebraicReference {
+            name,
+            poly_id: _,
+            next: false,
+        }) => Expression::Reference(
+            SourceRef::unknown(),
+            Reference::Poly(PolynomialReference {
+                name: name.clone(),
+                type_args: None,
+            }),
+        ),
+
+        AlgebraicExpression::BinaryOperation(AlgebraicBinaryOperation { left, op, right }) => {
+            BinaryOperation {
+                left: Box::new(
+                    try_algebraic_expression_to_expression(left)?
+                ),
+                op: (*op).into(),
+                right: Box::new(
+                    try_algebraic_expression_to_expression(right)?
+                ),
+            }.into()
+        }
+
+        AlgebraicExpression::UnaryOperation(AlgebraicUnaryOperation { op, expr }) => {
+            UnaryOperation {
+                op: (*op).into(),
+                expr: Box::new(
+                    try_algebraic_expression_to_expression(expr)?
+                ),
+            }.into()
+        }
+
+        AlgebraicExpression::Challenge(Challenge { id, stage }) => {
+            let function = Expression::Reference(
+                SourceRef::unknown(),
+                Reference::Poly(PolynomialReference {
+                    name: "std::prelude::challenge".to_string(),
+                    type_args: None,
+                }),
+            )
+            .into();
+            let arguments = [*stage as u64, *id]
+                .into_iter()
+                .map(|x| BigUint::from(x).into())
+                .collect();
+            Expression::FunctionCall(
+                SourceRef::unknown(),
+                FunctionCall {
+                    function,
+                    arguments,
+                },
+            )
+        }
+
+        AlgebraicExpression::Number(n) => Number {
+            value: n.to_arbitrary_integer(),
+            type_: Some(Type::Expr),
+        }
+        .into(),
+        _ => {
```

Comment:
References with `next` set to `true` like `x'` are also not covered by the previous cases, so the error message would be incorrect in that case.

Reply:
Maybe we shouldn't use `_` here so that we have a clear view of what is not implemented yet.

Reply:
Yes, and adding `todos` for the ones that the compiler complains about.

Reply:
Instead of matching value? Do you want me to return `Result<Expression>` instead of `Result<Expression, EvalError>`? and adding todo! macros?

Reply:
I just want to see what is missing. If it's just public reference, then let's implement it. It's just Expression::PublicReference.